app/scripts/controllers/decrypt-message.ts

```diff
@@ -0,0 +1,388 @@
+    this._keyringController = keyringController;
+    this._metricsEvent = metricsEvent;
+
+    this.hub = new EventEmitter();
+
+    this._decryptMessageManager = new DecryptMessageManager(
+      undefined,
+      undefined,
+      undefined,
+      ['decrypted'],
+    );
+
+    this._decryptMessageManager.hub.on('updateBadge', () => {
+      this.hub.emit('updateBadge');
+    });
+
+    this._decryptMessageManager.hub.on(
+      'unapprovedMessage',
+      (messageParams: AbstractMessageParamsMetamask) => {
+        this._requestApproval(messageParams);
+      },
+    );
+
+    this._subscribeToMessageState(
+      this._decryptMessageManager,
+      (state, newMessages, messageCount) => {
+        state.unapprovedDecryptMsgs = newMessages;
+        state.unapprovedDecryptMsgCount = messageCount;
+      },
+    );
+  }
+
+  /**
+   * A getter for the number of 'unapproved' Messages in the DecryptMessageManager.
+   *
+   * @returns The number of 'unapproved' Messages in the DecryptMessageManager.
+   */
+  get unapprovedDecryptMsgCount(): number {
+    return this._decryptMessageManager.getUnapprovedMessagesCount();
+  }
+
+  /**
+   * Reset the controller state to the initial state.
+   */
+  resetState() {
+    this.update(() => getDefaultState());
+  }
+
+  /**
+   * Clears all unapproved messages from memory.
+   */
+  clearUnapproved() {
+    this._decryptMessageManager.update({
+      unapprovedMessages: {},
+      unapprovedMessagesCount: 0,
+    });
+  }
+
+  /**
+   * Called when a dapp uses the eth_decrypt method
+   *
+   * @param messageParams - The params passed to eth_decrypt.
+   * @param req - The original request, containing the origin.
+   * @returns Promise resolving to the raw data of the signature request.
+   */
+  async newRequestDecryptMessage(
+    messageParams: DecryptMessageParams,
+    req: OriginalRequest,
+  ): Promise<string> {
+    return this._decryptMessageManager.addUnapprovedMessageAsync(
+      messageParams,
+      req,
+    );
+  }
+
+  /**
+   * Signifies a user's approval to decrypt a message in queue.
+   * Triggers decrypt, and the callback function from newUnsignedDecryptMessage.
+   *
+   * @param messageParams - The params of the message to decrypt & return to the Dapp.
+   * @returns A full state update.
+   */
+  async decryptMessage(messageParams: DecryptMessageParamsMetamask) {
+    const messageId = messageParams.metamaskId as string;
+    try {
+      const cleanMessageParams =
+        await this._decryptMessageManager.approveMessage(messageParams);
+
+      cleanMessageParams.data = this._parseMessageData(cleanMessageParams.data);
+      const rawMessage = await this._keyringController.decryptMessage(
+        cleanMessageParams,
+      );
+
+      this._decryptMessageManager.setMessageStatusAndResult(
+        messageId,
+        rawMessage,
+        'decrypted',
+      );
+      this._acceptApproval(messageId);
+    } catch (error) {
+      return this._cancelAbstractMessage(
```

Comment:
I'm unsure what's the correct behaviour in this case. But both `SignController` and `EncryptionPublicKeyController` throw when this happens. They also return the state within the try statement.

Not necessarily something to change, just an observation for now.

Reply:
Agree we should throw after cancelling and let the caller decide if they want to handle it.

Reply:
I've adjusted this part 👍